cvat-ui/src/utils/preview-queue.ts

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) CVAT.ai Corporation
+//
+// SPDX-License-Identifier: MIT
+
+import { Dispatch } from 'redux';
+import {
+    Job, Task, Project, CloudStorage,
+} from 'cvat-core-wrapper';
+import MLModel from 'cvat-core/src/ml-model';
+import { getJobPreviewAsync } from 'actions/jobs-actions';
+import { getTaskPreviewAsync } from 'actions/tasks-actions';
+import { getProjectsPreviewAsync } from 'actions/projects-actions';
+import { getCloudStoragePreviewAsync } from 'actions/cloud-storage-actions';
+import { getModelPreviewAsync } from 'actions/models-actions';
+
+export type PreviewEntity = Job | Task | Project | CloudStorage | MLModel;
+export type PreviewType = 'job' | 'task' | 'project' | 'cloudStorage' | 'model';
+
+export function getPreviewType(entity: PreviewEntity): PreviewType {
+    if (entity instanceof Job) return 'job';
+    if (entity instanceof Task) return 'task';
+    if (entity instanceof Project) return 'project';
+    if (entity instanceof CloudStorage) return 'cloudStorage';
+    if (entity instanceof MLModel) return 'model';
+    throw new Error('Unknown entity type');
+}
+
+export function getRequestId(entity: PreviewEntity): string {
+    const type = getPreviewType(entity);
+    return `${type}-${(entity as any).id}`;
+}
+
+interface PreviewRequest {
+    id: string;
+    entity: PreviewEntity;
+    dispatch: Dispatch;
+}
+
+const PREVIEW_REQUEST_DELAY = 100;
+
+class PreviewQueue {
+    private queue: PreviewRequest[] = [];
+    private isProcessing: boolean = false;
+    private currentRequest: PreviewRequest | null = null;
+
+    public addRequest(request: PreviewRequest): void {
+        const isDuplicate = this.queue.some((req) => req.id === request.id) ||
+                           (this.currentRequest && this.currentRequest.id === request.id);
+
+        if (!isDuplicate) {
+            this.queue.push(request);
+            this.processQueue();
+        }
+    }
+
+    public removeRequest(id: string): void {
+        this.queue = this.queue.filter((req) => req.id !== id);
+    }
+
+    private async processQueue(): Promise<void> {
+        if (this.isProcessing || this.queue.length === 0) {
+            return;
+        }
+
+        this.isProcessing = true;
+
+        while (this.queue.length > 0) {
+            const request = this.queue.shift()!;
+            this.currentRequest = request;
+
+            await this.executeRequest(request);
+
+            // Small delay between requests to avoid overwhelming the server
+            await new Promise<void>((resolve) => {
```

Comment:
Not sure we really need it. Just sending requests one by one is enough

Reply:
Removed.